### PR TITLE
PubMatic adapter

### PR DESCRIPTION
**Pull request by pm-harshad-mane**
## Type of change

- [ x] Bugfix
- [ ] Feature
- [ x] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ x] Build related changes
- [ x] CI related changes
- [ x] Other

## Description of change

- test parameters for validating bids
```
For a slot of 300x250: 
{
  bidder: 'pubmatic',
  params: {
    publisherId: '301', // MANDATORY
    adSlot: '/15671365/DMDemo@728x90:0', // MANDATORY
    pmzoneid: 'customZone1, customZone2', // OPTIONAL
    dctr: 'key1=value1,value2'  // OPTIONAL
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer: UOEDev@pubmatic.com
- [ ] official adapter submission

### CODE REVIEW

**Comment by dbemiller** [comment]
hey @pm-harshad-mane ... we're not accepting any pre-1.0 adapter PRs anymore. For more info, see [how to add a Prebid 1.0 Bidder Adapter](http://prebid.org/dev-docs/bidder-adapter-1.html). In particular, you'll need to use AJAX rather than JSONP, and go through the [bidderFactory](https://github.com/prebid/Prebid.js/blob/master/src/adapters/bidderFactory.js). See the AppNexusAst adapter (if you can request all the bids using a single request) or Rubicon adapter (if you need multiple requests to request all the bids) for examples

File: modules/pubmaticBidAdapter.js [hunk 1]
```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
...
+const constants = require('src/constants.json');
 
 /**
  * Adapter for requesting bids from Pubmatic.
  *
  * @returns {{callBids: _callBids}}
  * @constructor
  */
-function PubmaticAdapter() {
+var PubmaticAdapter = function PubmaticAdapter() {
```
Comment (mkendall07) [inline modules/pubmaticBidAdapter.js]: prefer `const`

File: modules/pubmaticBidAdapter.js [hunk 2]
```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
...
   var bids;
-  var _pm_pub_id;
-  var _pm_pub_age;
-  var _pm_pub_gender;
-  var _pm_pub_kvs;
-  var _pm_optimize_adslots = [];
+  var usersync = false;
```
Comment (mkendall07) [inline modules/pubmaticBidAdapter.js]: prefer `let`

File: modules/pubmaticBidAdapter.js [hunk 3]
```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
...
+      var bid = bids[i];
+      conf.pubId = conf.pubId || bid.params.publisherId;
+      conf = _handleCustomParams(bid.params, conf);
+      bid.params.adSlot = _cleanSlot(bid.params.adSlot);
+      bid.params.adSlot.length && slots.push(bid.params.adSlot);
+    }
+
+    // istanbul ignore else
+    if (conf.pubId && slots.length > 0) {
+      _legacyExecution(conf, slots);
```
Comment (mkendall07) [inline modules/pubmaticBidAdapter.js]: it looks like the legacy path is currently the only supported path (ie it still uses an iframe) is that correct? Is this optimized vs the last version?

File: modules/pubmaticBidAdapter.js [hunk 4]
```diff
@@ -2,137 +2,216 @@ var utils = require('src/utils.js');
...
+  function _cleanSlot(slotName) {
+    // istanbul ignore else
+    if (utils.isStr(slotName)) {
+      return slotName.replace(/^\s+/g, '').replace(/\s+$/g, '');
+    }
+    return '';
+  }
 
+  function _legacyExecution(conf, slots) {
+    var url = _generateLegacyCall(conf, slots);
```
Comment (mkendall07) [inline modules/pubmaticBidAdapter.js]: I'm not clear why this is still loaded into an iframe. Can you not use the `ajax` method here?

File: modules/pubmaticBidAdapter.js (hunk 3 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: Our older implementation used to include an external js library, now we do not include it.
_legacyExecution is the only flow available now, we are planning to add a new flow later.

File: modules/pubmaticBidAdapter.js (hunk 4 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: Our server response uses global namespace also our server is not supporting CORS.
By calling our server from a friendly-iframe we can get better priority than calling from a script tag thus we are calling our server using an iframe.

**Comment by pm-harshad-mane** [comment]
Hello @mkendall07 ,
Thank you for your review comments.
I have implemented your suggestions, also I have answered your queries.
Let me know if you need more information.

**Comment by mkendall07** [comment]
@pm-harshad-mane 
I'm not receiving bids. Here is the generated ad call:
`http://gads.pubmatic.com/AdServer/AdCallAggregator?SAVersion=1100&wp=PreBid&js=1&wv=prebid_prebid_0.33.0-pre&screenResolution=1440x900&ranreq=0.42388223356585986&inIframe=0&pageURL=http%3A%2F%2Flocalhost%3A9999%2FintegrationExamples%2F4_bidders.html&refurl=&kltstamp=2017-11-8%2015%3A37%3A40&timezone=-5&pubId=9999&kadpageurl=http%3A%2F%2Flocalhost%3A9999%2FintegrationExamples%2F4_bidders.html&dctr=key1%3Dvalue1%2Cvalue2&pmZoneId=customZone1%2C%20customZone2&adslots=%5B38519891%40300x250%5D`

Noticed that there was an `error: 2` in the response. Please advise.

**Comment by pm-harshad-mane** [comment]
Hello @mkendall07 ,

I have updated test publishe ID (301) and adSlot (/15671365/DMDemo@728x90:0), now you can see the response, you might need to try few times to get bid-responses.

Sample URI: http://gads.pubmatic.com/AdServer/AdCallAggregator?SAVersion=1100&wp=PreBid&js=1&wv=prebid_prebid_0.33.0-pre&screenResolution=1440x900&ranreq=0.42388223356585986&inIframe=0&pageURL=http%3A%2F%2Flocalhost%3A9999%2FintegrationExamples%2F4_bidders.html&refurl=&kltstamp=2017-11-8%2015%3A37%3A40&timezone=-5&pubId=301&kadpageurl=http%3A%2F%2Flocalhost%3A9999%2FintegrationExamples%2F4_bidders.html&dctr=key1%3Dvalue1%2Cvalue2&pmZoneId=customZone1%2C%20customZone2&adslots=%5B%2F15671365%2FDMDemo%40728x90%3A0%2C%2F15671365%2FDMDemo1%40300x250%3A0%2C%2F15671365%2FDMDemo2%40728x90%3A0%5D

**Comment by PubMatic-OpenWrap** [comment]
Thank you very much @mkendall07 :+1:

**Pull request merged**